core/src/main/java/com/scalar/db/storage/jdbc/JdbcService.java

```diff
@@ -182,21 +193,80 @@ public boolean mutate(
       Optional<String> namespace,
       Optional<String> tableName)
       throws SQLException, ExecutionException {
+    assert !connection.getAutoCommit();
     checkArgument(mutations.size() != 0);
     ScalarDbUtils.setTargetToIfNot(mutations, namespace, tableName);
     operationChecker.check(mutations);
 
-    for (Mutation mutation : mutations) {
-      if (mutation instanceof Put) {
-        if (!put((Put) mutation, connection, namespace, tableName)) {
-          return false;
-        }
-      } else {
-        if (!delete((Delete) mutation, connection, namespace, tableName)) {
-          return false;
+    Map<String, PreparedStatement> preparedStatementMap = new HashMap<>();
+    try {
+      for (Mutation mutation : mutations) {
+        if (mutation instanceof Put) {
+          if (!batchPut((Put) mutation, connection, preparedStatementMap)) {
+            return false;
+          }
+        } else {
+          if (!batchDelete((Delete) mutation, connection, preparedStatementMap)) {
+            return false;
+          }
         }
       }
+      for (PreparedStatement preparedStatement : preparedStatementMap.values()) {
+        preparedStatement.executeBatch();
+      }
+      return true;
+    } finally {
+      for (PreparedStatement preparedStatement : preparedStatementMap.values()) {
+        preparedStatement.close();
+      }
+    }
```

Comment:
This is the main change. But this optimization is only for mutations without any conditions, and we can pass mutations only for a single partition to the mutate API. So I wonder how many use cases can take advantage of this optimization.